### PR TITLE
DNS services are not resolving properly

### LINKED ISSUE #5154: Repeated skydns log: `incomplete CNAME chain: rcode is not equal to success`
Continuously:

```
2015-10-16 09:40:53.821119 I | skydns: incomplete CNAME chain: rcode is not equal to success
2015-10-16 09:41:08.831796 I | skydns: incomplete CNAME chain: rcode is not equal to success
2015-10-16 09:41:08.847089 I | skydns: incomplete CNAME chain: rcode is not equal to success
2015-10-16 09:41:23.810062 I | skydns: incomplete CNAME chain: rcode is not equal to success
2015-10-16 09:41:23.817699 I | skydns: incomplete CNAME chain: rcode is not equal to success
2015-10-16 09:41:38.807745 I | skydns: incomplete CNAME chain: rcode is not equal to success
2015-10-16 09:41:38.836402 I | skydns: incomplete CNAME chain: rcode is not equal to success
2015-10-16 09:41:53.817832 I | skydns: incomplete CNAME chain: rcode is not equal to success
2015-10-16 09:41:53.820592 I | skydns: incomplete CNAME chain: rcode is not equal to success
2015-10-16 09:42:08.815018 I | skydns: incomplete CNAME chain: rcode is not equal to success
2015-10-16 09:42:08.817311 I | skydns: incomplete CNAME chain: rcode is not equal to success
2015-10-16 09:42:23.841105 I | skydns: incomplete CNAME chain: rcode is not equal to success
2015-10-16 09:42:23.844855 I | skydns: incomplete CNAME chain: rcode is not equal to success
```

### PR DESCRIPTION
**Pull request by smarterclayton**
The change on Sept 15th changed how services resolved in the absence
of search paths, which resulted in very long times to resolve DNS in
some cases.

This reduces search times to a few ms.

Fixes #5154 

@liggitt

### CODE REVIEW

**Comment by smarterclayton** [comment]
[test][extended:core]

**Comment by smarterclayton** [comment]
Fixes #5154

**Comment by liggitt** [comment]
Add tests to end to end to make sure short names still resolve? We currently have this:
```
MASTER_SERVICE_IP="$(dig @${API_HOST} "kubernetes.default.svc.cluster.local." +short A | head -n 1)"
```
Also test these?
```
kubernetes.default.svc
kubernetes.default
kubernetes
```

**Comment by liggitt** [comment]
actually, not sure shortnames will work unless we have search paths set up

**Comment by smarterclayton** [comment]
dns_test is pretty good.  It's also very tolerant.  I'd like to wait until
we have the SCC stuff to enable the upstream e2e.

**Comment by smarterclayton** [comment]
v1.0.6:

```
$ dig @localhost _endpoints.kubernetes.default.svc.cluster.local

; <<>> DiG 9.9.6-P1-RedHat-9.9.6-8.P1.fc21 <<>> @localhost _endpoints.kubernetes.default.svc.cluster.local
; (2 servers found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 14752
;; flags: qr aa rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 0

;; QUESTION SECTION:
;_endpoints.kubernetes.default.svc.cluster.local. IN A

;; ANSWER SECTION:
_endpoints.kubernetes.default.svc.cluster.local. 30 IN A 10.0.2.15

;; Query time: 86 msec
;; SERVER: 127.0.0.1#53(127.0.0.1)
;; WHEN: Tue Nov 03 02:11:01 UTC 2015
;; MSG SIZE  rcvd: 81

[vagrant@openshiftdev origin]$ dig @localhost _endpoints.kubernetes.default.svc.cluster.local SRV

; <<>> DiG 9.9.6-P1-RedHat-9.9.6-8.P1.fc21 <<>> @localhost _endpoints.kubernetes.default.svc.cluster.local SRV
; (2 servers found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 39167
;; flags: qr aa rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; QUESTION SECTION:
;_endpoints.kubernetes.default.svc.cluster.local. IN SRV

;; ANSWER SECTION:
_endpoints.kubernetes.default.svc.cluster.local. 30 IN SRV 10 50 8443 unknown-port-8443.e1.kubernetes.

;; ADDITIONAL SECTION:
unknown-port-8443.e1.kubernetes. 30 IN	A	10.0.2.15

;; Query time: 253 msec
;; SERVER: 127.0.0.1#53(127.0.0.1)
;; WHEN: Tue Nov 03 02:11:08 UTC 2015
;; MSG SIZE  rcvd: 163
[vagrant@openshiftdev origin]$ dig @localhost kubernetes.default.svc.cluster.local

; <<>> DiG 9.9.6-P1-RedHat-9.9.6-8.P1.fc21 <<>> @localhost kubernetes.default.svc.cluster.local
; (2 servers found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 27725
;; flags: qr aa rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 0

;; QUESTION SECTION:
;kubernetes.default.svc.cluster.local. IN A

;; ANSWER SECTION:
kubernetes.default.svc.cluster.local. 30 IN A	172.30.0.1

;; Query time: 133 msec
;; SERVER: 127.0.0.1#53(127.0.0.1)
;; WHEN: Tue Nov 03 02:12:26 UTC 2015
;; MSG SIZE  rcvd: 70

[vagrant@openshiftdev origin]$ dig @localhost kubernetes.default.svc.cluster.local SRV

; <<>> DiG 9.9.6-P1-RedHat-9.9.6-8.P1.fc21 <<>> @localhost kubernetes.default.svc.cluster.local SRV
; (2 servers found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 56804
;; flags: qr aa rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; QUESTION SECTION:
;kubernetes.default.svc.cluster.local. IN SRV

;; ANSWER SECTION:
kubernetes.default.svc.cluster.local. 30 IN SRV	10 50 443 unknown-port-443.portal.kubernetes.

;; ADDITIONAL SECTION:
unknown-port-443.portal.kubernetes. 30 IN A	172.30.0.1

;; Query time: 113 msec
;; SERVER: 127.0.0.1#53(127.0.0.1)
;; WHEN: Tue Nov 03 02:12:29 UTC 2015
;; MSG SIZE  rcvd: 158
```

File: pkg/dns/serviceresolver.go [hunk 1]
```diff
@@ -72,8 +76,30 @@ func (b *ServiceResolver) Records(name string, exact bool) ([]msg.Service, error
 	if len(segments) == 0 {
 		return nil, nil
 	}
+	glog.V(4).Infof("Answering query %s:%t", dnsName, exact)
+	switch base := segments[0]; base {
+	case "pod":
```
Comment (liggitt) [inline pkg/dns/serviceresolver.go]: where did this come from?

Comment (smarterclayton) [inline pkg/dns/serviceresolver.go]: kube added it

File: test/integration/dns_test.go [hunk 2]
```diff
@@ -175,11 +185,11 @@ func TestDNS(t *testing.T) {
 			dnsQuestionName: "headless2.default.svc.cluster.local.",
 			srv: []*dns.SRV{
 				{
-					Target: "http.e1.headless2.",
+					Target: headless2IPHash + "._http._tcp.headless2.default.svc.cluster.local.",
 					Port:   2346,
 				},
 				{
-					Target: "other.e1.headless2.",
+					Target: headless2IPHash + "._other._tcp.headless2.default.svc.cluster.local.",
```
Comment (liggitt) [inline test/integration/dns_test.go]: citation needed

File: pkg/dns/serviceresolver.go [hunk 3]
```diff
@@ -60,11 +64,11 @@ func NewServiceResolver(config *server.Config, accessor ServiceAccessor, endpoin
 //   * endpoint_id is "portal" when portalIP is set
 // * endpoints always returns each individual endpoint as A records
```
Comment (liggitt) [inline pkg/dns/serviceresolver.go]: update godoc with changes to returned records and description of pod dns name and returned records

File: pkg/dns/serviceresolver.go (hunk 1 above)
Comment (liggitt) [inline pkg/dns/serviceresolver.go]: pointer to their stuff so we can be sure we're consistent?

File: pkg/dns/serviceresolver.go [hunk 4]
```diff
@@ -278,3 +286,27 @@ func extractIP(reverseName string) (string, bool) {
...
+}
+
+// return a hash for the key name
+func getHash(text string) string {
+	h := fnv.New32a()
+	h.Write([]byte(text))
+	return fmt.Sprintf("%x", h.Sum32())
+}
+
+func reverseIP(ip string) string {
```
Comment (liggitt) [inline pkg/dns/serviceresolver.go]: doc or rename, I expected this to reverse the string

File: pkg/dns/serviceresolver.go (hunk 1 above)
Comment (smarterclayton) [inline pkg/dns/serviceresolver.go]: https://github.com/kubernetes/kubernetes/blob/master/cluster/addons/dns/README.md

**Comment by smarterclayton** [comment]
Any other comments?

**Comment by smarterclayton** [comment]
SRV records changed from `_proto._port.SVCNAME` to `_proto._port.SVCNAME.NAMESPACE.svc.cluster.local` (should resolve the same, but faster)

**Comment by smarterclayton** [comment]
Endpoint SRV records change from `_port.EP<NUM>.SVCNAME` to `ENDPOINTIPHASH._port._proto.SVCNAME.NAMESPACE.svc.cluster.local`

**Comment by smarterclayton** [comment]
[merge]

**Pull request merged**